File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const dummy = new ListNode(-1);
    let list1Current = list1;
    let list2Current = list2;
    let current = dummy;
    while (list1Current !== null && list2Current !== null) {
        if (list1Current.val < list2Current.val) {
            current.next = list1Current;
            list1Current = list1Current.next;
        } else {
            current.next = list2Current;
            list2Current = list2Current.next;
        }
        current = current.next;
    }
    if (list1Current !== null) {
        current.next = list1Current;
    }
    if (list2Current !== null) {
        current.next = list2Current;
    }
    return dummy.next;
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    result := &ListNode{}
    tail := result

    for list1 != nil && list2 != nil {
        if list1.Val < list2.Val {
            tail.Next = &ListNode{
                Val: list1.Val,
                Next: &ListNode{},
            }
            list1 = list1.Next
        } else {
            tail.Next = &ListNode{
                Val: list2.Val,
                Next: &ListNode{},
            }
            list2 = list2.Next
        }

        tail = tail.Next
    }

    if list1 != nil {
        tail.Next = list1
    } else if list2 != nil {
        tail.Next = list2
    }

    return result.Next
}